llama-index-integrations/llms/llama-index-llms-ollama/llama_index/llms/ollama/base.py

```diff
@@ -107,329 +167,239 @@ def _model_kwargs(self) -> Dict[str, Any]:
             **self.additional_kwargs,
         }
 
+    def _convert_to_ollama_messages(self, messages: Sequence[ChatMessage]) -> Dict:
+        return [
+            {
+                "role": message.role.value,
+                "content": message.content or "",
+            }
+            for message in messages
+        ]
+
+    def _prepare_chat_with_tools(
+        self,
+        tools: List["BaseTool"],
+        user_msg: Optional[Union[str, ChatMessage]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        verbose: bool = False,
+        allow_parallel_tool_calls: bool = False,
+        **kwargs: Any,
+    ) -> Dict[str, Any]:
+        tool_specs = [
+            tool.metadata.to_openai_tool(skip_length_check=True) for tool in tools
+        ]
+
+        if isinstance(user_msg, str):
+            user_msg = ChatMessage(role=MessageRole.USER, content=user_msg)
+
+        messages = chat_history or []
+        if user_msg:
+            messages.append(user_msg)
+
+        return {
+            "messages": messages,
+            "tools": tool_specs or None,
+        }
+
+    def _validate_chat_with_tools_response(
+        self,
+        response: ChatResponse,
+        tools: List["BaseTool"],
+        allow_parallel_tool_calls: bool = False,
+        **kwargs: Any,
+    ) -> ChatResponse:
+        """Validate the response from chat_with_tools."""
+        if not allow_parallel_tool_calls:
+            force_single_tool_call(response)
+        return response
+
+    def get_tool_calls_from_response(
+        self,
+        response: "ChatResponse",
+        error_on_no_tool_call: bool = True,
+    ) -> List[ToolSelection]:
+        """Predict and call the tool."""
+        tool_calls = response.message.additional_kwargs.get("tool_calls", [])
+
+        if len(tool_calls) < 1:
+            if error_on_no_tool_call:
+                raise ValueError(
+                    f"Expected at least one tool call, but got {len(tool_calls)} tool calls."
+                )
+            else:
+                return []
+
+        tool_selections = []
+        for tool_call in tool_calls:
+            argument_dict = tool_call["function"]["arguments"]
```

Comment:
does this work for streaming atm?

Reply:
Hard to say, I didn't test streaming extensively yet 🥲 I suspect it might just insert the entire tool call on the last delta

Reply:
Or it doesn't support it at all for streaming

Reply:
yeah that's fine, mistral ai also returns the entire tool call in its "streaming" endpoint, just checking there's no bugs if you run a streaming code snippet

Reply:
Confirmed with Ollama that streaming + tool calls does not work currently (it won't ever return a tool call)